### DISCUSSION TITLE
How to implement layered architecture (Service Layer + Data Layer) with actix and diesel

### DISCUSSION THREAD

**Discussion by 0xDones** [actix/actix-web#2670] [Q&A]
Hi guys, I'm a beginner to Rust and I'm struggling a lot trying to design a layered application using some patterns like having `Services` for business logic and `Repositories` for data access, I would love some help to understand what I'm doing wrong.

So far I have:

- `user/controller.rs` 
  - UserController -> Handles HTTP
- `user/service.rs` 
  - UserService -> Business Logic
- `user/repository.rs` 
  - UserRepository -> Database Interactions
- `user/model.rs` 
  - User -> Entity

`UserService` and `UserRepository` are `structs` with methods. `UserService` has an instance of a `UserRepository`, which has an instance of a database pool manager. I'm instantiating all those things in my `main` function and wrapping the `UserService` using ` web::Data` and passing it to the `Application State`, in order to access it inside my `request handlers`, but I simply can't get it to work.

I put all the source code in this repo, with all the instructions to start the app: https://github.com/refl3ction/rust-api

I'm currently stuck in the `UserController::create_user` function, in which I can't get the `UserService` from Application state.

Appreciate some help 🙏

**Comment by Akronae** [comment]
Looking for that feature too

**Reply by Johnmuchiri** [reply]
https://github.com/microsoft/cookiecutter-rust-actix-clean-architecture/blob/main/README.md

**Comment by Samielakkad** [comment]
A common Actix shape for this is:

```text
handler/controller -> service -> repository -> database pool/connection
```

The important bit is that `web::Data<T>` stores application state by type, so the type you insert into `App::app_data(...)` must exactly match the type you extract in the handler.

For example, I would usually make the service cloneable and store it once:

```rust
#[derive(Clone)]
pub struct UserRepository {
    pool: DbPool,
}

impl UserRepository {
    pub fn new(pool: DbPool) -> Self {
        Self { pool }
    }
}

#[derive(Clone)]
pub struct UserService {
    repo: UserRepository,
}

impl UserService {
    pub fn new(repo: UserRepository) -> Self {
        Self { repo }
    }

    pub async fn create_user(&self, input: CreateUser) -> Result<User, AppError> {
        self.repo.create(input).await
    }
}
```

Then register it like this:

```rust
let pool = create_pool();
let user_repo = UserRepository::new(pool.clone());
let user_service = UserService::new(user_repo);

HttpServer::new(move || {
    App::new()
        .app_data(web::Data::new(user_service.clone()))
        .configure(user_routes)
})
```

And extract the exact same type in the handler:

```rust
pub async fn create_user(
    service: web::Data<UserService>,
    payload: web::Json<CreateUser>,
) -> Result<HttpResponse, Error> {
    let user = service
        .create_user(payload.into_inner())
        .await
        .map_err(error::ErrorInternalServerError)?;

    Ok(HttpResponse::Created().json(user))
}
```

A couple of gotchas:

- Do not create a new repository/service inside every handler. Put shared dependencies in `web::Data`.
- Do not store a single database connection as app state. Store a pool, then get a connection per operation/request.
- If your service contains non-`Clone` internals, wrap those internals in `Arc`, or store `web::Data<Arc<UserService>>`, but do not double-wrap accidentally unless you extract the same type.
- Keep HTTP things out of the repository. The repository should not know about `HttpResponse`, `web::Json`, etc.

For Diesel specifically, the exact repository implementation depends on whether you use sync Diesel with `web::block` / a pool, or `diesel_async`. But the Actix app-state pattern is the same: build the pool once, build the service once from that pool/repository, put the service into `web::Data`, and extract `web::Data<UserService>` in handlers.

If you are seeing ?app data is not configured? or extractor failures, it almost always means the inserted type and extracted type do not match exactly, or the `.app_data(...)` was added to a different scope/app than the route that needs it.
